### DISCUSSION TITLE
Force components to be allocated with specific alignment / custom allocator?

### DISCUSSION THREAD

**Discussion by jonesmz** [skypjack/entt#666] [Q&A]
This is a preliminary question, and not something that's fully fleshed out.

Anyway, I have a situation where I want to use SIMD on some of the data stored in EnTT, which requires the data be aligned to a specific alignment size. Currently, I'm copying that data out of EnTT, and into a manually allocated buffer, doing my algorithm, and then copying the data back into EnTT.

I'm curious if EnTT has pre-existing support for indicating data alignment, or just providing a custom allocator, so that I (might) skip some of that data copying.

A related, but not exactly the same, situation is that I may want to communicate some data that is stored in EnTT to a GPU as a single buffer. Are there any facilities that EnTT offers that I should use to do that?

**Comment by skypjack** [comment]
You can use custom pools here. That is, you can replace the _storage part_ of a pool on a per-type basis. It should solve your problem, am I wrong?
As for the single buffer, there is the `raw` and `size` pair from the basic pool. Don't know if it fits here, but it seems so:
```
auto view = registry.view<T>();
auto size = view.size(); // number of elements
auto *raw = view.raw(); // pointer to first element
```

**Reply by qwiff-dogg** [reply]
Like the OP, I'd like to manage my component's memory using a custom allocator. I have a hard time understanding how to implement this. I'm trying to achieve this using the new Allocator mechanism in v3.8.

From what I can understand, you have to provide `entt::storage_traits` and `std::allocator_traits` for this to work. 

```cpp

class CustomAllocator {
// allocate() and deallocate() omitted
}

namespace entt {

    template<typename Entity, typename Type>
    struct storage_traits<Entity, Type> {
        using storage_type = sigh_storage_mixin<basic_storage<Entity, Type, CustomAllocator>>;
    };

}

namespace std {

    struct allocator_traits<CustomAllocator> {

        template<typename U>
        using rebind_alloc = std::allocator<U>;

        template<typename U>
        using rebind_traits = allocator_traits<std::allocator<U>>;
    };

}
```

This compiles and runs but simply maps the allocation for `Entity` and `Type` to `std::allocator`. Any hints on how to proceed from here? Thanks in advance.

**Reply by skypjack** [reply]
The long term plan is to make the allocator leak to the registry so as to set a registry-wide allocator and stop using the `storage_traits` for that. Until then, yeah, this is the way to go.
From here, all you need is to implement your allocator properly. This is more of a C++ question than an EnTT question.
You don't really have to specialize `std::allocator_traits` as long as your allocator makes all the required aliases available. See [here](https://en.cppreference.com/w/cpp/memory/allocator_traits) for further details.

**Reply by qwiff-dogg** [reply]
@skypjack Thanks for your quick reply. It's good to hear that are more custom allocation features on the roadmap.

I've implemented a "by the numbers" C++ allocator, which seems to work.

One snag I encountered is that I have to override the [default Allocator template](https://github.com/skypjack/entt/blob/2c35203647677c8330eefd5ec1d7a4ba2f862ba4/src/entt/entity/fwd.hpp#L12-L13) argument for `basic_sparse_set` in order for `basic_registry` to compile with a custom `storage_traits` with an allocator other than `std::allocator<Type>`.

The only way to provide a new `Allocator` default is by using a replacement fwd header for `entt/entity/fwd.hpp`. Is my assumption correct or is there a more elegant way? I realize it's more a C++ question. Templates are not my strongest suit, tbh.

**Reply by skypjack** [reply]
Hi, I think I know the culprit for that. The following should be enough:
```cpp
template<typename Entity, typename Type>
struct entt::storage_traits<Entity, Type> {
    using storage_type = sigh_storage_mixin<basic_storage<Entity, Type, CustomAllocator>>;
};
```
However, if you're using v3.8.x, it's likely that it doesn't work because of the view class.
It should be fixed already on `master`. Briefly, I didn't expect users to replace the allocator until it reaches the registry and therefore I missed some fundamental changes to make it work without overriding the stuff from `fwd.hpp`.
It would be great if you could make a test with `master` and let me know if it works without overriding the forward declaration.

**Reply by qwiff-dogg** [reply]
`master` doesn't seem to make a difference.

Compiling with clang produces the same error as before:

`(117, 30) Cannot initialize a parameter of type 'std::__1::unique_ptr<entt::basic_sparse_set<entt::entity>>::pointer' (aka 'entt::basic_sparse_set<entt::entity> *') with an rvalue of type 'storage_type<game::Environment> *' (aka 'entt::sigh_storage_mixin<entt::basic_storage_impl<entt::entity, game::Environment, CustomAllocator<game::Environment>>> *')`

This is due to the use of the [basic_sparse_set](https://github.com/skypjack/entt/blob/1639429edd0f962e9c5f47e6498efbbd34e0d16a/src/entt/entity/registry.hpp#L48) template alias. `basic_sparse_set` should declare the same allocator type as specified in `entt::storage_traits`.

But as you said, custom allocation is not fully implemented yet. I'll just have to wait until it is. In the meantime I'll rely on overriding the fwd header.

**Reply by skypjack** [reply]
Oh, yeah, that's because I've updated the view but I've not updated the registry yet. You're right. 😅 Sorry for the noise.

**Reply by qwiff-dogg** [reply]
@skypjack Is there any news regarding the use of custom allocators with a `registry` in 3.9 and 3.10? I'm looking at the changelogs and can't really tell.

**Reply by skypjack** [reply]
Versions 3.9 and 3.10 added allocator (and `uses_allocator` support) to sparse sets, storage, dense map, dense set, dispatcher, locator, resource handling, and so on.
The registry isn't in this list yet for a few reasons:
* Time 🤷‍♂️ 
* Other priorities besides this 🤷‍♂️ 
* I want to put entities in their own storage to avoid doing the work to support allocator twice (since the former is planned anyway)

However, there are a few ways already to inject and use your allocator with a registry. For example, if it's default constructible you can just use the `storage_traits` and make it the default:
  ```cpp
  template<typename Entity, typename Type>
  struct storage_traits<Entity, Type> {
      using storage_type = sigh_storage_mixin<basic_storage<Entity, Type, my_allocator>>;
  };
  ```
Otherwise, you can add a mixin that finds the right allocator somewhere and injects it to the base class. These aren't the only options but they should give you a grasp of what I mean at least.

**Reply by qwiff-dogg** [reply]
@skypjack Thanks for the reply. I finally had some time to look into this. 

For 3.10 I use the same solution as I did with 3.8.

1. Provide my own `storage_traits` referencing the custom allocator type.
2. Include my own version of `entt/registry/fwd.hpp` with default template args for [`basic_sparse_set` and `basic_storage`](https://github.com/skypjack/entt/blob/1639429edd0f962e9c5f47e6498efbbd34e0d16a/src/entt/entity/fwd.hpp#L12-L17) referencing the custom allocator type, instead of `std::allocator`. 

Without step 2), you'll get the compile errors I mentioned in my earlier reply.

**Reply by skypjack** [reply]
Oh, yeah, the registry isn't ready yet in this sense. Granted. Work in progress.
So far, only a registry-less model works perfectly fine. In all other cases, you've to trick the type resolution.
The problem with the registry is that all pools are lazily initialized. Therefore, we'll have to provide the allocator type explicitly.
But yeah, just a matter of time, I'm working on this stuff too among the other things.
